### PR TITLE
Let tuner auto reconnect to NNI manager

### PR DESCRIPTION
**Pull request by liuzhe-lz**
### Description ###

Dispatcher now tries to reconnect when it loses connection to NNI manager.
For one command it only retries once. If it fails twice for the same message then dispatcher crashes.
Because the reconnect is raised by client (dispatcher) side, dispatcher to manager messages will be resent on connection loss, but manager to dispatcher messages during the reconnect (~5s) will loss. I think it's okay because it only hurts tuner's accuracy.

Because connection error is now "expected", we add a new tuner command `ER` (error) to report real tuner error.
When the outermost function of dispatcher catches an exception, it tries to send the content with error command.
In NNI manager side, the shim layer of IPC channel is modified to report `ER` content to NNI manager and hide connection errors.

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [x] test case
  - ~~doc~~

### How to test ###

I don't know...

### CODE REVIEW

File: nni/__main__.py [hunk 1]
```diff
@@ -65,6 +66,7 @@ def main():
         tuner._on_error()
         if assessor is not None:
             assessor._on_error()
+        dispatcher.report_error(traceback.format_exc())
```
Comment (ultmaster) [inline nni/__main__.py:69]: If web socket already loses connection, will it end up here? What will `report_error` behave?

File: nni/runtime/tuner_command_channel/channel.py [hunk 2]
```diff
@@ -51,10 +52,19 @@ def disconnect(self) -> None:
...
+        except WebSocket.ConnectionClosed:
+            self._channel = WebSocket(self._url)
+            self._channel.send(command)
 
     def _receive(self) -> tuple[CommandType, str] | tuple[None, None]:
-        command = self._channel.receive()
+        try:
+            command = self._channel.receive()
+        except WebSocket.ConnectionClosed:
+            self._channel = WebSocket(self._url)
```
Comment (ultmaster) [inline nni/runtime/tuner_command_channel/channel.py:65]: We'd better log this as a warning. Having more logs doesn't hurt.

Do you think retrying immediately and for only once is a good idea? What's the reason for this?

File: nni/runtime/tuner_command_channel/websocket.py [hunk 3]
```diff
@@ -85,10 +88,10 @@ def receive(self) -> str | None:
             msg = _wait(self._ws.recv())
             _logger.debug(f'Received {msg}')
         except websockets.ConnectionClosed:  # type: ignore
-            _logger.debug('Connection closed by server.')
+            _logger.warning('Connection closed by server.')
```
Comment (ultmaster) [inline nni/runtime/tuner_command_channel/websocket.py:91]: Why is this only in receive? How about send?

**Review by QuanluZhang — APPROVED** [review]

File: nni/runtime/tuner_command_channel/channel.py [hunk 4]
```diff
@@ -55,17 +61,45 @@ def _send(self, command_type: CommandType, data: str) -> None:
...
         if command is None:
-            raise RuntimeError('NNI manager closed connection')
+            command = self._retry_receive()
         command_type = CommandType(command[:2].encode())
         return command_type, command[2:]
+
+    def _retry_receive(self) -> str:
+        _logger.warning('Connection lost. Trying to reconnect...')
+        for interval in self._retry_intervals:
+            time.sleep(interval)
```
Comment (matluster) [inline nni/runtime/tuner_command_channel/channel.py:95]: Suggest adding logs at every retrying.

**Review by ultmaster — APPROVED** [review]

**Pull request merged**